vhdbuilder/packer/install-dependencies.sh

```diff
@@ -395,11 +396,96 @@ for KUBE_PROXY_IMAGE_VERSION in ${KUBE_PROXY_IMAGE_VERSIONS}; do
   CONTAINER_IMAGE="mcr.microsoft.com/oss/kubernetes/kube-proxy:v${KUBE_PROXY_IMAGE_VERSION}"
   pullContainerImage ${cliTool} ${CONTAINER_IMAGE}
   ctr --namespace k8s.io run --rm ${CONTAINER_IMAGE} checkTask /bin/sh -c "iptables --version" | grep -v nf_tables && echo "kube-proxy contains no nf_tables"
-  
+
   # shellcheck disable=SC2181
   echo "  - ${CONTAINER_IMAGE}" >>${VHD_LOGS_FILEPATH}
 done
 
+# download and setup azcopy to use to download private packages with MSI auth
+getAzCopyCurrentPath() {
+  if [[ -f ./azcopy ]]; then
+    echo "./azcopy already exists"
+  else
+    echo "get azcopy at \"${PWD}\"...start"
+    # Download and extract
+    azcopydownloadurl="https://aka.ms/downloadazcopy-v10-linux"
+    if [[ $(isARM64) == 1 ]]; then
+      azcopydownloadurl="https://aka.ms/downloadazcopy-v10-linux-arm64"
+    fi
+    wget "$azcopydownloadurl" -O "downloadazcopy"
+    tar -xvf ./downloadazcopy
+
+    rm -f ./azcopy
+    cp ./azcopy_linux_*/azcopy ./azcopy
+    chmod +x ./azcopy
+
+    rm -f downloadazcopy
+    rm -rf ./azcopy_linux_*/
+    echo "get azcopy...done"
+  fi
+}
+
+# download the given tgz package and extract/import the container image
+importContainerImageTGZ() {
+  CONTAINER_IMAGE_TGZ=$1
+
+  if out=$(ctr --namespace k8s.io image import ${CONTAINER_IMAGE_TGZ}); then # ctr should be there by default now, don't have to handle crictl/docker tools
+    container_image=$(echo $out | cut -d' ' -f 2)
+    echo "imported image $container_image from $CONTAINER_IMAGE_TGZ"
+    retagImageForAllClouds "$CLI_TOOL" "$container_image"
+  else
+    echo "error importing container image from $CONTAINER_IMAGE_TGZ"
+    exit 1
+  fi
+}
+
+# download kubernetes package from the given URL using MSI for auth for azcopy
+# if it is a kube-proxy package, extract image from the downloaded package
+cacheKubePackageFromPrivateUrl() {
+  KUBE_PRIVATE_BINARY_URL=$1
+  IS_KUBE_PROXY_PKG=$2
+
+  echo "process private package url: $KUBE_PRIVATE_BINARY_URL"
+
+  mkdir -p ${K8S_CACHE_DIR} # /opt/kubernetes/downloads/private-packages
+
+  if [[ "$IS_KUBE_PROXY_PKG" == true ]]; then
+    # save kube-proxy pkg with the same/given package name
+    K8S_TGZ_NAME=${KUBE_PRIVATE_BINARY_URL##*/}
+  else
+    # save kube pkg with version number from the url path, this convention is used to find the cached package at run-time
+    K8S_TGZ_NAME=$(echo "$KUBE_PRIVATE_BINARY_URL" | grep -o -P '(?<=\/kubernetes\/).*(?=\/binaries\/)').tar.gz
+  fi
+
+  # use azcopy with MSI instead of curl to download packages
+  getAzCopyCurrentPath
+
+  export AZCOPY_AUTO_LOGIN_TYPE="MSI"
+  export AZCOPY_MSI_RESOURCE_STRING="$LINUX_MSI_RESOURCE_IDS"
```

Comment:
I wonder if we actually need to specify the linux MSI here, I'd think azcopy would be smart enough to infer the resource ID of the MSI that's actually attached to the VM

Reply:
ah, I guess we do since this should be a user-assigned MSI...

https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-authorize-azure-active-directory#authorize-by-using-a-user-assigned-managed-identity